translatable_shared/src/translations/node.rs

```diff
@@ -0,0 +1,146 @@
+use std::collections::HashMap;
+
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{ToTokens, TokenStreamExt, quote};
+use strum::ParseError;
+use thiserror::Error;
+use toml::{Table, Value};
+
+use crate::macros::collections::map_transform_to_tokens;
+use crate::misc::language::Language;
+use crate::misc::templating::{TemplateError, validate_format_string};
+
+/// Errors occurring during TOML-to-translation structure transformation
+#[derive(Error, Debug)]
+pub enum TranslationNodeError {
+    /// Mixed content found in nesting node (strings and objects cannot coexist)
+    #[error("A nesting can contain either strings or other nestings, but not both.")]
+    InvalidNesting,
+
+    /// Template syntax error with unbalanced braces
+    #[error("Tempalte validation failed: {0:#}")]
+    UnclosedTemplate(#[from] TemplateError),
+
+    /// Invalid value type encountered in translation structure
+    #[error("Only strings and objects are allowed for nested objects.")]
+    InvalidValue,
+
+    /// Failed to parse language code from translation key
+    #[error("Couldn't parse ISO 639-1 string for translation key")]
+    LanguageParsing(#[from] ParseError),
+}
+
+pub type TranslationNesting = HashMap<String, TranslationNode>;
+pub type TranslationObject = HashMap<Language, String>;
+
+/// Represents nested translation structure,
+/// as it is on the translation files.
+#[derive(Clone)]
+pub enum TranslationNode {
+    /// Nested namespace containing other translation objects
+    Nesting(TranslationNesting),
+    /// Leaf node containing actual translations per language
+    Translation(TranslationObject),
+}
+
+impl TranslationNode {
+    /// Resolves a translation path through the nesting hierarchy
+    ///
+    /// # Arguments
+    /// * `path` - Slice of path segments to resolve
+    ///
+    /// # Returns
+    /// Reference to translations if path exists and points to leaf node
+    pub fn find_path<I: ToString>(&self, path: &Vec<I>) -> Option<&TranslationObject> {
```

Comment:
What if, instead of passing a `Vec` here, we changed it to a `&[I]`? Then we wouldn't need to convert the slice back to a vector down here.

Reply:
This can't be done, I need a heap allocation because keeping a reference alive for so long would add too much overhead, and also possibly require to add lifetimes to structures who possibly don't need it.

So in other words, it can be done, but there are implications to it. In some or other way I would be heap allocating it.

Reply:
Makes sense, well just ignore the others comments about this one.